ink_linting/src/mapping_initialized.rs

```diff
@@ -0,0 +1,345 @@
+) -> Option<Span> {
+    let mut visitor = InkAttributeVisitor {
+        cx,
+        ink_attribute: None,
+        constructor_info: None,
+    };
+
+    walk_item(&mut visitor, item);
+
+    match visitor.constructor_info {
+        Some(info) if !info.uses_initialize_contract => Some(info.span),
+        _ => None,
+    }
+}
+
+/// Visitor for ink! attributes.
+struct InkAttributeVisitor<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    ink_attribute: Option<InkAttribute>,
+    constructor_info: Option<InkConstructor>,
+}
+
+// Information about an ink! constructor.
+struct InkConstructor {
+    // The constructor has a call to `ink_lang::utils::initialize_contract(…)`
+    // in its function.
+    uses_initialize_contract: bool,
+    // The span for the constructor function.
+    span: Span,
+}
+
+impl<'tcx> Visitor<'tcx> for InkAttributeVisitor<'_, 'tcx> {
+    type Map = Map<'tcx>;
+
+    fn visit_fn(
+        &mut self,
+        kind: FnKind<'tcx>,
+        decl: &'tcx FnDecl<'_>,
+        body_id: BodyId,
+        span: Span,
+        id: HirId,
+    ) {
+        // We can return immediately if an incorrect constructor was already found
+        if let Some(constructor) = &self.constructor_info {
+            if !constructor.uses_initialize_contract {
+                return;
+            }
+        }
+
+        let attrs = self.cx.tcx.get_attrs(id.owner.to_def_id());
+        self.ink_attribute = get_ink_attribute(attrs);
+
+        if self.ink_attribute == Some(InkAttribute::Storage) {
+            return;
+        } else if self.ink_attribute == Some(InkAttribute::Constructor) {
+            log::debug!("Found constructor, starting to search for `initialize_contract`");
+            let mut visitor = InitializeContractVisitor {
+                cx: self.cx,
+                uses_initialize_contract: false,
+            };
+            walk_fn(&mut visitor, kind, decl, body_id, span, id);
+
+            log::debug!(
+                "Has `initialize_contract`? {:?}",
+                visitor.uses_initialize_contract
+            );
+            self.constructor_info = Some(InkConstructor {
+                uses_initialize_contract: visitor.uses_initialize_contract,
+                span,
+            });
+
+            return;
+        }
+
+        walk_fn(self, kind, decl, body_id, span, id);
+    }
+
+    fn nested_visit_map(&mut self) -> NestedVisitorMap<Self::Map> {
+        NestedVisitorMap::All(self.cx.tcx.hir())
+    }
+}
+
+/// Visitor to determine if a `fn` contains a call to `ink_lang::utils::initialize_contract(…)`.
+///
+/// # Known Limitation
+///
+/// This function currently only finds call to `initialize_contract` which happen
+/// on the first level of the `fn` ‒ no nested calls are found! So if you would
+/// call `initialize_contract` within a sub-function of the ink! constructor
+/// this is not recognized!
+///
+/// Also the function call must be explicitly called `ink_lang::utils::initialize_contract(…)`,
```

Comment:
Is this true though? I modified the `mapping_one_constructor` to `use ink_lang::utils::initialize_contract;` and it still works. Also if you look at the regex it uses the non reexported name to match on